manualdirectdebit.php

```diff
@@ -166,3 +166,147 @@ function manualdirectdebit_civicrm_navigationMenu(&$menu) {
     _manualdirectdebit_civix_insert_navigation_menu($menu, 'Administer/' . $directDebitMenuItem['name'], $menuItem);
   }
 }
+
+/**
+ * Implements hook_civicrm_postProcess().
+ *
+ */
+function manualdirectdebit_civicrm_postProcess($formName, &$form) {
+  $action = $form->getAction();
+
+  if ($formName == 'CRM_Contribute_Form_Contribution' && $action == CRM_Core_Action::ADD) {
+    $contributionId = (int) $form->getVar('_id');
+    $assignedContactId = (int) $form->getVar('_contactID');
+    $mandateId = getIdOfNewMandate($assignedContactId);
+    if ($form->_params['is_recur'] == 1) {
+      $recurrContributionId = $form->_params['contributionRecurID'];
+      saveDependencyBetweenContributionAndMandate($recurrContributionId, $mandateId);
+    }
+
+    saveMandateInContribution($contributionId, $mandateId);
+  }
+}
+
+/**
+ * Gets id of last inserted direct debit mandate
+ *
+ * @param $assignedContactId
+ *
+ * @return int
+ */
+function getIdOfNewMandate($assignedContactId) {
+  $tableName = 'civicrm_value_dd_mandate';
+
+  createNewMandate($assignedContactId, $tableName);
+
```

Comment:
We suppose that in CiviCRM, exist more clear approach how to get last inserted ID. But API, and BAO/DAO don`t give the expected result.

Reply:
unfortunately it appear that civi API does not achieve what we want and it is hard to use BAO/DAO in this case, and yeah the method you used is not reliable and is subject to race condition, though you can just assign the result of the insert query to a variable then use ->fetch() method to get the ID instead of using LAST_INSERT_ID().

Reply:
We try to use ->fetch(), as you mentioned, but 'DB_resultid' always return null, and we suggest to use script like this "CRM_Core_DAO::executeQuery("SELECT MAX(`id`) AS id FROM `civicrm_value_dd_mandate` WHERE `entity_id` = $assignedContactId"); " for getting last inserted id.

Reply:
I will try to do some local testing and see what can be done, but I really would like to avoid max(id), it is not an ideal solution.

Reply:
Please note that query returns MAX(id) for particular entity_id - *WHERE entity_id = $assignedContactId");*. 

1. So if two users simultaneously creates two mandates but for different entity_id we will get correct record.  
2. But it could return wrong id if two users creates two mandates for the same entity_id 

To be make code work correctly for any cases  - we should add dummy column to the table, code will INSERT unique "generated key" into this field. And next query will return assigned mandate_id for this "generated key".  

Like *SELECT id FROM Table WHERE generated-key ='98-76-5-49'*

Reply:
Yeah I did some testing and research on this and this is a fine solution, the suggestion of using generated key is a good one too, though I am not sure if it is possible to create two mandates for the same user at the same time but I don't think this is the case so we can ignore the generated key and just stick to this solution, thanks @AWG-Bohdan  and @awg-github